internal/controller/ocirepository_controller.go

```diff
@@ -689,11 +694,90 @@ func (r *OCIRepositoryReconciler) verifySignature(ctx context.Context, obj *ociv
 		}
 
 		return fmt.Errorf("no matching signatures were found for '%s'", ref)
+
+	case "notation":
+		// get the public keys from the given secret
+		secretRef := obj.Spec.Verify.SecretRef
+
+		if secretRef == nil {
+			return fmt.Errorf("verification secret cannot be empty: '%s'", ref)
+		}
+
+		verifySecret := types.NamespacedName{
+			Namespace: obj.Namespace,
+			Name:      secretRef.Name,
+		}
+
+		pubSecret, err := r.retrieveSecret(ctxTimeout, verifySecret)
+		if err != nil {
+			return err
+		}
+
+		data, ok := pubSecret.Data[notation.DefaultTrustPolicyKey]
+		if !ok {
+			return fmt.Errorf("'%s' not found in secret '%s'", notation.DefaultTrustPolicyKey, verifySecret.String())
+		}
+
+		var doc trustpolicy.Document
+
+		if err := json.Unmarshal(data, &doc); err != nil {
+			return err
+		}
+
+		signatureVerified := false
+
+		defaultNotationOciOpts := []notation.Options{
+			notation.WithTrustStore(&doc),
+			notation.WithRemoteOptions(opt...),
+			notation.WithAuth(auth),
+			notation.WithKeychain(keychain),
+			notation.WithInsecureRegistry(obj.Spec.Insecure),
+		}
+
+		for k, data := range pubSecret.Data {
+			if strings.HasSuffix(k, ".crt") || strings.HasSuffix(k, ".pem") {
+				verifier, err := notation.NewNotationVerifier(append(
+					defaultNotationOciOpts,
+					notation.WithPublicCertificate(data))...)
+				if err != nil {
+					return err
+				}
+
+				signatures, err := verifier.Verify(ctxTimeout, ref)
+				if err != nil {
+					r.eventLogf(ctx, obj, corev1.EventTypeWarning, sourcev1.VerificationError,
```

Comment:
Love it! A couple notes.

- In your example you are appending the error to an `[]error{}`. However, only one `error` is returned if the array is greater than `0`. What do you think about using `return errors.Join(sigRetErr...)` instead?
- For `helmchart_controller` when verifying the chart, the `error` is return straight away in the form of a new `error` here https://github.com/fluxcd/source-controller/blob/main/internal/helm/repository/oci_chart_repository.go#L384 (slightly different to how `ocirepository_controller` works). To keep things consistent, perhaps we could return `sigRetErr` wrapped with the same message? Something like `return fmt.Errorf("failed to verify %s: %w", ref, errors.Join(sigRetErr...))`

Reply:
> Let's update the OCIRepository notation verification code to not ignore any error when encountered. I think that'll take care of ErrorSignatureRetrievalFailed and any other possible failures.

No worries. Resolved in [66e3e1e](https://github.com/fluxcd/source-controller/pull/1075/commits/66e3e1ef59712e2d41f14eb85120ae14d5aea587).

> Similarly, the audit level, as per https://github.com/notaryproject/specifications/blob/main/specs/trust-store-trust-policy.md#signature-verification-details, IIUC, only actually verifies if a signature is present.

That's my understanding as well. `audit` checks the validity of the signature and ensures the artefact has not been tampered with after signing. If the signature is invalid because the artefact had been altered after signing, validation should fail too. However, if the signature is not present (artefact was not signed), or the signature is not authentic, then it would be logged instead. 

In those situations, the user has decided based on their trust policy they are okay with allowing the artefact to deploy. I don't think we should do anything different like returning an error to prevent the artefact from deploying. I think at the very least we should ensure those audits are bubbling their way to the surface and logged within flux and rely on notation's verification process + trust policy.

What I will do is add additional unit tests to cover a range of different verification types and update the verifier to handle those types better so that the results are better represented in the logs. 

Would that work for you?

Reply:
The `cosign` verification would analog to `strict` in the notation doc. It believe we should point the user to this information somehow (even though one would think they know about it).

I think `Verify()` should be used in all our controllers, but we I wouldn't change the `cosign` code in the pull request. `Verify` should handle all the verification logic include needed checks and surface errors accordingly. The controllers should not have to interpret `notation` errors imo. I think returning early in case of error is good here, we shouldn't give too much information.

Reply:
> However, if the signature is not present (artefact was not signed), or the signature is not authentic, then it would be logged instead.

This is not an accurate characterization of how audit mode is described in the table under this heading:
https://github.com/notaryproject/specifications/blob/main/specs/trust-store-trust-policy.md#signature-verification-details

Audit mode will fail if there is no signature to verify, because integrity is checked and enforced. Permissive mode will fail additionally if the signature is not authenticated, but other types of failure are only logged and do not block verification. Skip mode will never check for signatures and will not verify anything. It does not make sense for Flux to report a verification success condition IMHO. Strict is enforcing all the checks including time-based and expiry.

So I think there is a question about how Flux should signal in each of these cases, since the notary notation does not have exactly the same semantics as cosign. The cosign verification isn't modal in the same way - keyed or keyful mode, and keyless mode. Signatures are always verified (there is no "audit" or "skip" model afaik). Additional constraints can be placed on the verification with the regex-based issuer and subject matching, but there is no analogue to the permissive or audit mode, and "skip" mode would be just disable verification. So the signaling should be different with notary model. We had some discussion about this at Bug Scrub today, just writing up my thoughts in reflection – I'm not sure if this is responsive to the thread

Reply:
If you think in terms of what will be the response whenever the signature fails, it makes a lot more sense to think in terms of the notary model. On further reflection to think about the modes in terms of adoption and rolling back. I had to step into thinking like a user that wants to adopt signed artifact verification and what obstacles I may face.

If verification fails, and you don't have an permissive/audit or skip mode, then the first time an incident is created because of some signature verification that was due to a transient failure or other surprise, the valid only response that's possible is to disable verification and not enable it again, until the issue is confirmed to be resolved. This is what I think you call a total systems failure - the protection is removed because it can't be used effectively. Because there is no permissive fallback mode, verification can't be performed, even in a limited fashion. (And that's bad...)

You can (with the notary model) put it into permissive mode, which ignores all time based failures, or audit mode which doesn't do any authentication of the sources against a specific key at all. Only validates signatures haven't been tampered with. This is similar to keyless mode of cosign, but with additional fallback options. This way you can think of adopting the signature verification in a progressive model, which you can't do with the cosign `verify` configuration. It's either enabled or it isn't. There is no "skip" configuration. It is questionable if Flux should honor a similar "skip" configuration because from a validation perspective, you don't really want a configuration that skips validation once you've committed to enable validation. But as a user, you may want to progressively enable the validation when you are first adopting if we assume many validation failures won't manifest as the result of a real attack. We don't want to introduce failures into the system, except in a controlled way. We can progressively enable the notary notation as we progress from first publishing signatures, then validating and/or authenticating, then upgrading from permissive into strict mode, gaining increasing levels of confidence about the artifact, in a way that isn't currently possible with the cosign config OCI source spec.verify section so far as I can tell.

So when you first start signing releases, and you want to verify the signatures, if you find that you have problems because of hard to narrow down issues like server times are not all set to match each other, or something weird like that, you can avoid most of those failures and still check the authenticity - this maybe opens you up to something like a reply attack, but you're still verifying that the expected key has signed the artifact, even if there are issues. This is progress. And it's a better response posture, you can respond to that incident that some surprising thing resulted in timing issues that caused a verification to fail - not by disabling source authentication, but by making it permissive. That sounds useful! Later on, when the NTPD issues are resolved, you can re-enable strict mode. On your incident report, you can note that at no time were unverified sources required to be used in the system.

Also positive, vs. "we had to disable verification as long as we couldn't access rekor" - again not sure how responsive these notes are to the thread, but I wanted to capture this from the discussion we had about it at Bug Scrub. These are things to consider with the new implementation.

Reply:
> This doesn't seem like what we discussed and agreed upon above. It is my initial implementation I was explicitly looking for `ErrorSignatureRetrievalFailed` error.

Ah right. Apologies @darkowlzz. I miss understood what you meant. 

> I think we need to return the verification result based on these values...

Just so we are on the same page, do you want `VerificationOutcome` returned to the controller to allow for logging if `err == nil` and `len(signatures) > 0`, else just return a `false and an `err`? Or do you want a special error type as you suggest above and let the controller handle how it is logged? Another alternative, we could create a `logr.Logger` option and allow for `ctrl.LoggerFrom(ctx)` to be passed in, then from there we could write the logs from within `notation.Verify()` prior to returning the results.

Reply:
@darkowlzz I've made the updates in [6cb0253](https://github.com/fluxcd/source-controller/pull/1075/commits/6cb0253c4dc22a8573f069a93d8b5f77e70d6e1a). I have not yet updated `helmchart_controller` to avoid marking `VerificationResultIgnored` as `true`. I'll sort that out later in the day.

Reply:
From the last discussion I understood we would optionally log the different `logged` validations.

Something like:

```go
...
	_, outcomes, err := notation.Verify(ctx, *v.verifier, repo, verifyOptions)
	if err != nil {
		return oci.VerificationResultFailed, err
	}

	if len(outcomes) == 0 {
		return oci.VerificationResultFailed, fmt.Errorf("signature verification failed for all the signatures associated with %s", url)
	}

	outcome := outcomes[0]

	switch outcome.VerificationLevel {
	case trustpolicy.LevelSkip:
		return oci.VerificationResultIgnored, nil
	case trustpolicy.LevelStrict:
		return oci.VerificationResultSuccess, nil
	case trustpolicy.LevelPermissive:
		// Optionally log Authentic timestamp, and optionals Expiry and Revocation checks
		return oci.VerificationResultSuccess, nil
	case trustpolicy.LevelAudit:
		// Optionally log Authenticity, Authentic timestamp, and optionals Expiry and Revocation checks
		return oci.VerificationResultSuccess, nil
	default:
		return oci.VerificationResultFailed, errors.New("unknown verification level")
	}
}
```

Reply:
With regards to logging, are you okay with leaving a comment mentioning _Optional log here..._ just as you suggest in your example, @souleb, or do you want me to implement logging? I could have an option to pass in `logr.Logger` from the controller and use that to log the different outcomes.

Reply:
Yes, an option to pass a logger is fine I think.

> Originally, I did implement the same switch method as you but opted against it as notation does allow you to [override the various actions](https://github.com/notaryproject/notation-go/blob/main/verifier/trustpolicy/trustpolicy.go#L169).

If I read the code correctly, here:
https://github.com/notaryproject/notation-go/blob/7fa8404e79d95dfb8ff0a89754dcdc57cea6945b/verifier/verifier.go#L147 we get a `verificationLevel` that is `custom` with the `overrides` applied. If correct, then the `outcome` used in the switch should be correct as is.